File: src/libs/ThirdParty.VBSuppliers/Properties/GoGlobal/GoGlobal.vb
Imports System.Xml
Imports Intuitive
Imports Intuitive.Net.WebRequests
Imports Intuitive.Functions
Imports ThirdParty.Abstractions
Imports ThirdParty.Abstractions.Models.Property.Booking
Imports ThirdParty.Abstractions.Models
Imports ThirdParty.Abstractions.Constants
Imports ThirdParty.Abstractions.Lookups
Imports System.Text.RegularExpressions
Imports Intuitive.Helpers.Extensions
Imports System.Net.Http

Public Class GoGlobal
    Implements IThirdParty

#Region "Constructor"

    Public Sub New(settings As IGoGlobalSettings, support As ITPSupport, httpClient As HttpClient)
        _settings = Ensure.IsNotNull(settings, NameOf(settings))
        _support = Ensure.IsNotNull(support, NameOf(support))
        _httpClient = Ensure.IsNotNull(httpClient, NameOf(httpClient))
    End Sub

#End Region

#Region "Properties"

    Private ReadOnly _settings As IGoGlobalSettings

    Private ReadOnly _support As ITPSupport

    Private ReadOnly _httpClient As HttpClient

    Public ReadOnly Property SupportsBookingSearch As Boolean Implements IThirdParty.SupportsBookingSearch
        Get
            Return True
        End Get
    End Property

    Private Function SupportsLiveCancellation(searchDetails As IThirdPartyAttributeSearch, source As String) As Boolean Implements IThirdParty.SupportsLiveCancellation
        Return _settings.AllowCancellations(searchDetails)
    End Function

    Public ReadOnly Property SupportsRemarks As Boolean Implements IThirdParty.SupportsRemarks
        Get
            Return True
        End Get
    End Property
    Private Function TakeSavingFromCommissionMargin(searchDetails As IThirdPartyAttributeSearch) As Boolean Implements IThirdParty.TakeSavingFromCommissionMargin
        Return False
    End Function

    Private Function OffsetCancellationDays(searchDetails As IThirdPartyAttributeSearch) As Integer Implements IThirdParty.OffsetCancellationDays
        Return _settings.OffsetCancellationDays(searchDetails, False)
    End Function

    Public Function RequiresVCard(info As VirtualCardInfo) As Boolean Implements IThirdParty.RequiresVCard
        Return False
    End Function

    Public ReadOnly Property PreBookRequestNumber As Integer
        Get
            Return 9
        End Get
    End Property
    Public ReadOnly Property BookRequestNumber As Integer
        Get
            Return 2
        End Get
    End Property
    Public ReadOnly Property StatusConfirmationRequestNumber As Integer
        Get
            Return 5
        End Get
    End Property
    Public ReadOnly Property CancelRequestNumber As Integer
        Get
            Return 3
        End Get
    End Property
    Public ReadOnly Property CancellationCostRequestNumber As Integer
        Get
            Return 4
        End Get
    End Property

#End Region

#Region "Prebook"

    Public Function PreBook(oPropertyDetails As PropertyDetails) As Boolean Implements IThirdParty.PreBook

        Dim bReturn As Boolean = True

        Try
            Dim xPreBook As XDocument = SendRequest("Pre-Book", PreBookRequestNumber, ValuationRequest(oPropertyDetails, _settings), oPropertyDetails, _settings)
            bReturn = CostsAndCancellation(oPropertyDetails, xPreBook)
            oPropertyDetails.LocalCost = oPropertyDetails.Rooms.Sum(Function(r) r.LocalCost)

        Catch ex As Exception

            oPropertyDetails.Logs.AddNew(ThirdParties.GOGLOBAL, "PreBook Exception", ex.ToString)
            bReturn = False

        End Try

        Return bReturn

    End Function

#End Region

#Region "Book"

    Public Function Book(oPropertyDetails As PropertyDetails) As String Implements IThirdParty.Book

        Dim sReference As String = ""

        Try
            Dim xBook As XDocument = SendRequest("Book", BookRequestNumber, BookReservationRequest(oPropertyDetails, _settings), oPropertyDetails, _settings)
            sReference = xBook.Document.Element("Main").Element("GoBookingCode").Value

            Dim xStatus As XDocument = SendRequest("Status Confirmation", StatusConfirmationRequestNumber,
                                                   StatusRequest(oPropertyDetails, sReference, _settings), oPropertyDetails, _settings)

            If Not xStatus.Document.Element("Main").Element("GoBookingCode").Attribute("Status").Value.InList("C", "VCH") Then
                oPropertyDetails.Logs.AddNew(ThirdParties.GOGLOBAL, "Third Party / Book Failed", "Booking unsuccessful")

                sReference = "Failed"
            End If

        Catch ex As Exception

            oPropertyDetails.Logs.AddNew(ThirdParties.GOGLOBAL, "Book Exception", ex.ToString)
            sReference = "Failed"

        End Try
        Return sReference
    End Function

#End Region

#Region "Cancel"

    Public Function CancelBooking(oPropertyDetails As PropertyDetails) As ThirdPartyCancellationResponse Implements IThirdParty.CancelBooking

        Dim oThirdPartyCancellationResponse As New ThirdPartyCancellationResponse

        Try
            Dim oXML As XDocument = SendRequest("Cancel", CancelRequestNumber, CancelRequest(oPropertyDetails, _settings), oPropertyDetails, _settings)

        Catch ex As Exception

            oPropertyDetails.Logs.AddNew(ThirdParties.GOGLOBAL, "Cancellation Exception", ex.ToString)
            oThirdPartyCancellationResponse.TPCancellationReference = "Failed"

        End Try

        Return oThirdPartyCancellationResponse

    End Function

    Public Function GetCancellationCost(oPropertyDetails As PropertyDetails) As ThirdPartyCancellationFeeResult Implements IThirdParty.GetCancellationCost

        Dim oThirdPartyCancellationFeeResult As New ThirdPartyCancellationFeeResult

        Try
            Dim oXML As XDocument = SendRequest("Cancellation Cost", CancellationCostRequestNumber, ViewReservationRequest(oPropertyDetails), oPropertyDetails, _settings)

        Catch ex As Exception

            oPropertyDetails.Logs.AddNew(ThirdParties.GOGLOBAL, "Cancellation Cost Exception", ex.ToString)
            oThirdPartyCancellationFeeResult.Success = False

        End Try

        Return oThirdPartyCancellationFeeResult

    End Function

#End Region

#Region "BookingSearch"

    Public Function BookingSearch(oBookingSearchDetails As BookingSearchDetails) As ThirdPartyBookingSearchResults Implements IThirdParty.BookingSearch

        Dim oThirdPartyBookingSearchResults As New ThirdPartyBookingSearchResults

        Return oThirdPartyBookingSearchResults

    End Function

#End Region

#Region "Other Methods"

    Public Function CreateReconciliationReference(sInputReference As String) As String Implements IThirdParty.CreateReconciliationReference
        Return ""
    End Function

#Region "Booking Status Update"

    Public Function BookingStatusUpdate(ByVal oPropertyDetails As PropertyDetails) As ThirdPartyBookingStatusUpdateResult Implements IThirdParty.BookingStatusUpdate
        Return New ThirdPartyBookingStatusUpdateResult
    End Function

#End Region

#End Region
    Public Sub EndSession(oPropertyDetails As PropertyDetails) Implements IThirdParty.EndSession

    End Sub



#Region "Helpers"

    Private Shared Function StripInvalidCharacters(ByVal Line As String) As String

        Return Regex.Replace(Line, "[^A-Za-z- ]", String.Empty).ToProperCase()

    End Function

    Private Function SendRequest(ByVal RequestType As String, ByVal RequestNumber As Integer, ByVal XML As String,
                                        ByVal PropertyDetails As PropertyDetails, ByVal Settings As IGoGlobalSettings) As XDocument

        Dim xsi As XNamespace = XNamespace.Get("http://www.w3.org/2001/XMLSchema-instance")
        Dim xsd As XNamespace = XNamespace.Get("http://www.w3.org/2001/XMLSchema")
        Dim soap As XNamespace = XNamespace.Get("http://schemas.xmlsoap.org/soap/envelope/")
        Dim ns As XNamespace = XNamespace.Get("http://www.goglobal.travel/")

        Dim xRequest As XDocument = New XDocument(
          New XDeclaration("1.0", "utf-8", "no"),
          New XElement(soap + "Envelope",
           New XAttribute(XNamespace.Xmlns + "xsi", xsi),
           New XAttribute(XNamespace.Xmlns + "xsd", xsd),
           New XAttribute(XNamespace.Xmlns + "soap", soap),
           New XElement(soap + "Body",
         New XElement(ns + "MakeRequest",
          New XAttribute("xmlns", ns),
          New XElement(ns + "requestType", RequestNumber),
          New XElement(ns + "xmlRequest", New XCData(XML))))))

        Dim oRequest As New Request
        With oRequest
            .EndPoint = Settings.URL(PropertyDetails)
            .SetRequest(xRequest.ToString)
            .Method = eRequestMethod.POST
            .Source = ThirdParties.GOGLOBAL
            .LogFileName = RequestType
            .CreateLog = True
            .UseGZip = True
            .Send(_httpClient)
        End With

        If oRequest.RequestLog <> "" Then
            PropertyDetails.Logs.AddNew(ThirdParties.GOGLOBAL, "GoGlobal " & RequestType & " Request", oRequest.RequestLog)
        End If

        If oRequest.ResponseLog <> "" Then
            PropertyDetails.Logs.AddNew(ThirdParties.GOGLOBAL, "GoGlobal " & RequestType & " Response", oRequest.ResponseLog)
        End If

        Dim XResponse As XDocument = XDocument.Parse(oRequest.ResponseString)

        Dim xInfo As XDocument = XDocument.Parse(XResponse.Document.Element(soap + "Envelope").Element(soap + "Body").Element(ns + "MakeRequestResponse").Element(ns + "MakeRequestResult").Value)
        Dim xHotel As XElement = xInfo.Document.Element("Root").Element("Main")

        Return XDocument.Parse(xHotel.ToString)

    End Function

    Private Shared Function ValuationRequest(ByVal PropertyDetails As PropertyDetails, ByVal Settings As IGoGlobalSettings) As String

        Dim iSalesChannelID As Integer = PropertyDetails.SalesChannelID
        Dim iBrandID As Integer = PropertyDetails.BrandID

        Dim xRequest As XDocument = New XDocument(New XElement("Root",
                        New XElement("Header",
                         New XElement("Agency", Settings.Agency(PropertyDetails)),
                         New XElement("User", Settings.User(PropertyDetails)),
                         New XElement("Password", Settings.Password(PropertyDetails)),
                         New XElement("Operation", "BOOKING_VALUATION_REQUEST"),
                         New XElement("OperationType", "Request")
                         ),
                      New XElement("Main",
                          New XElement("HotelSearchCode", PropertyDetails.Rooms(0).ThirdPartyReference.Split("|"c)(0)),
                          New XElement("ArrivalDate", PropertyDetails.ArrivalDate.ToString("yyyy-MM-dd")))))

        Return xRequest.ToString

    End Function

    Private Shared Function BookReservationRequest(ByVal PropertyDetails As PropertyDetails, ByVal Settings As IGoGlobalSettings) As String

        Dim iSalesChannelID As Integer = PropertyDetails.SalesChannelID
        Dim iBrandID As Integer = PropertyDetails.BrandID
        Dim sGuestID As Integer = 1
        For Each oPassenger As Passenger In PropertyDetails.Rooms(0).Passengers
            oPassenger.TPID = sGuestID.ToString()
            sGuestID += 1
        Next

        Dim xRequest As XDocument = New XDocument(New XElement("Root",
                        New XElement("Header",
                         New XElement("Agency", Settings.Agency(PropertyDetails)),
                         New XElement("User", Settings.User(PropertyDetails)),
                         New XElement("Password", Settings.Password(PropertyDetails)),
                         New XElement("Operation", "BOOKING_INSERT_REQUEST"),
                         New XElement("OperationType", "Request")
                         ),
                      New XElement("Main",
                          New XElement("AgentReference", PropertyDetails.BookingReference),
                          New XElement("HotelSearchCode", PropertyDetails.Rooms(0).ThirdPartyReference.Split("|"c)(0)),
                          New XElement("ArrivalDate", PropertyDetails.ArrivalDate.ToString("yyyy-MM-dd")),
                          New XElement("Nights", PropertyDetails.Duration),
                          New XElement("NoAlternativeHotel", 1),
                          New XElement("Leader",
                              New XAttribute("LeaderPersonID", 1)),
                          New XElement("Rooms",
                              From oRoom In PropertyDetails.Rooms
                              Select New XElement("RoomType",
                                        New XAttribute("Adults", oRoom.Adults + oRoom.ChildAges.Where(Function(i) i > 10).Count),
                                       If(oRoom.Infants > 0, New XAttribute("Cots", oRoom.Infants), Nothing),
                                       New XElement("Room",
                                       New XAttribute("RoomID", -1 * oRoom.PropertyRoomBookingID),
                                       From oGuest In oRoom.Passengers.Where(Function(o) o.PassengerType.ToString <> "Infant")
                                       Select New XElement(If(oGuest.PassengerType.ToString = "Adult" OrElse oGuest.Age > 10, "PersonName", "ExtraBed"),
                                            StripInvalidCharacters(String.Format("{0} {1} {2}", oGuest.FirstName, oGuest.LastName, oGuest.Title)),
                                            New XAttribute("PersonID", oGuest.TPID),
                                            If(oGuest.PassengerType.ToString = "Child" AndAlso oGuest.Age < 11, New XAttribute("ChildAge", oGuest.Age), Nothing))))))))

        Return xRequest.ToString

    End Function

    Private Function ViewReservationRequest(ByVal PropertyDetails As PropertyDetails) As String

        Dim iSalesChannelID As Integer = PropertyDetails.SalesChannelID
        Dim iBrandID As Integer = PropertyDetails.BrandID

        Dim xRequest As XDocument = New XDocument(New XElement("Root",
                        New XElement("Header",
                         New XElement("Agency", _settings.Agency(PropertyDetails)),
                         New XElement("User", _settings.User(PropertyDetails)),
                         New XElement("Password", _settings.Password(PropertyDetails)),
                         New XElement("Operation", "BOOKING_SEARCH_REQUEST"),
                         New XElement("OperationType", "Request")
                         ),
                      New XElement("Main",
                          New XElement("GoBookingCode", PropertyDetails.SourceReference))))

        Return xRequest.ToString

    End Function

    Private Shared Function VoucherIssueRequest(ByVal PropertyDetails As PropertyDetails, ByVal sSourceReference As String, ByVal Settings As IGoGlobalSettings) As String

        Dim iSalesChannelID As Integer = PropertyDetails.SalesChannelID
        Dim iBrandID As Integer = PropertyDetails.BrandID

        Dim xRequest As XDocument = New XDocument(New XElement("Root",
                        New XElement("Header",
                         New XElement("Agency", Settings.Agency(PropertyDetails)),
                         New XElement("User", Settings.User(PropertyDetails)),
                         New XElement("Password", Settings.Password(PropertyDetails)),
                         New XElement("Operation", "VOUCHER_DETAILS_REQUEST"),
                         New XElement("OperationType", "Request")
                         ),
                      New XElement("Main",
                          New XElement("GoBookingCode", sSourceReference))))

        Return xRequest.ToString

    End Function

    Private Shared Function StatusRequest(ByVal PropertyDetails As PropertyDetails, ByVal sSourceReference As String, ByVal Settings As IGoGlobalSettings) As String

        Dim iSalesChannelID As Integer = PropertyDetails.SalesChannelID
        Dim iBrandID As Integer = PropertyDetails.BrandID

        Dim xRequest As XDocument = New XDocument(New XElement("Root",
                        New XElement("Header",
                         New XElement("Agency", Settings.Agency(PropertyDetails)),
                         New XElement("User", Settings.User(PropertyDetails)),
                         New XElement("Password", Settings.Password(PropertyDetails)),
                         New XElement("Operation", "BOOKING_STATUS_REQUEST"),
                         New XElement("OperationType", "Request")
                         ),
                      New XElement("Main",
                          New XElement("GoBookingCode", sSourceReference))))

        Return xRequest.ToString

    End Function

    Private Shared Function CancelRequest(ByVal PropertyDetails As PropertyDetails, ByVal Settings As IGoGlobalSettings) As String

        Dim iSalesChannelID As Integer = PropertyDetails.SalesChannelID
        Dim iBrandID As Integer = PropertyDetails.BrandID

        Dim xRequest As XDocument = New XDocument(New XElement("Root",
                        New XElement("Header",
                         New XElement("Agency", Settings.Agency(PropertyDetails)),
                         New XElement("User", Settings.User(PropertyDetails)),
                         New XElement("Password", Settings.Password(PropertyDetails)),
                         New XElement("Operation", "BOOKING_CANCEL_REQUEST"),
                         New XElement("OperationType", "Request")
                         ),
                      New XElement("Main",
                          New XElement("GoBookingCode", PropertyDetails.SourceReference))))

        Return xRequest.ToString

    End Function

    Private Shared Function CostsAndCancellation(ByVal PropertyDetails As PropertyDetails, ByVal xDoc As XDocument) As Boolean

        Dim bAvailable As Boolean = True
        Dim oCancellationsList As New Cancellations
        Dim oCancellation As New Cancellation

        Try

            'Have added the remarks recieved at search into the TPReference
            For Each oRoom As RoomDetails In PropertyDetails.Rooms
                If oRoom.ThirdPartyReference.Split("|"c).Count > 2 AndAlso oRoom.ThirdPartyReference.Split("|"c)(2) <> "" Then
                    PropertyDetails.Errata.AddNew("", oRoom.ThirdPartyReference.Split("|"c)(2))
                End If
            Next

            'Get the cost
            Dim xRates As XElement = xDoc.Document.Element("Main").Element("Rates")
            Dim nRates As Decimal = PropertyDetails.LocalCost
            If xRates IsNot Nothing AndAlso xRates.Value <> "" Then
                nRates = Regex.Match(xRates.Value, "\d+(?:.\d+)?").ToString.ToSafeDecimal()

                If nRates > 0 AndAlso nRates <> PropertyDetails.Rooms(0).LocalCost Then
                    For Each oRoom As RoomDetails In PropertyDetails.Rooms
                        oRoom.LocalCost = nRates / PropertyDetails.Rooms.Count
                        oRoom.GrossCost = nRates / PropertyDetails.Rooms.Count
                    Next
                End If
            End If

            oCancellation.Amount = nRates
            Dim sDeadline As String = xDoc.Document.Element("Main").Element("CancellationDeadline").Value

            'Now cancellations/errata
            Dim xRemarks As XElement = xDoc.Document.Element("Main").Element("Remarks")
            If xRemarks IsNot Nothing AndAlso xRemarks.Value <> "" Then
                Dim sErrata As String = xRemarks.Value

                'Check for cancelation stuff in errata
                Dim rMatch As Match = Regex.Match(sErrata, "(?:STARTING )([0-9]+[- /.][0-9][0-9][- /.][0-9]+)(?: CXL-PENALTY FEE IS )(?:PRICE OF )?([0-9.]+)(?:%| NIGHTS)")
                If rMatch IsNot Nothing AndAlso rMatch.Success Then

                    sDeadline = rMatch.Groups(1).ToString
                    Dim nValue As Decimal = rMatch.Groups(2).ToString.ToSafeDecimal()

                    'First try percentage
                    If rMatch.ToString.Contains("%") Then
                        oCancellation.Amount = nRates * (nValue / 100).ToSafeMoney()
                    Else
                        oCancellation.Amount = nRates * (nValue / PropertyDetails.Duration).ToSafeMoney()
                    End If

                End If

                PropertyDetails.Errata.AddNew("", sErrata)

            End If

            oCancellation.StartDate = sDeadline.ToSafeDate()
            oCancellation.EndDate = New Date(2099, 12, 31)

        Catch ex As Exception
            bAvailable = False

            'If it fails, assume it's non-refundable
            oCancellation.Amount = PropertyDetails.LocalCost
            oCancellation.StartDate = Date.Now.AddDays(-1)

        End Try

        oCancellationsList.Add(oCancellation)

        PropertyDetails.Cancellations = Cancellations.MergeMultipleCancellationPolicies(oCancellationsList)

        Return bAvailable

    End Function

#End Region

End Class

File: src/libs/iVectorOne.Suppliers/GoGlobal/GoGlobal.cs
namespace iVectorOne.Suppliers.GoGlobal
{
    using System;
    using System.Collections.Generic;
    using System.Globalization;
    using System.Linq;
    using System.Net.Http;
    using System.Text.RegularExpressions;
    using System.Threading.Tasks;
    using Intuitive;
    using Intuitive.Helpers.Extensions;
    using Intuitive.Helpers.Net;
    using Intuitive.Helpers.Serialization;
    using iVectorOne.Constants;
    using iVectorOne.Interfaces;
    using iVectorOne.Models;
    using iVectorOne.Models.Property;
    using iVectorOne.Models.Property.Booking;
    using iVectorOne.Suppliers.GoGlobal.Models;
    using Microsoft.Extensions.Logging;

    public class GoGlobal : IThirdParty, ISingleSource
    {
        #region "Properties"

        private readonly IGoGlobalSettings _settings;
        private readonly ISerializer _serializer;
        private readonly HttpClient _httpClient;
        private readonly ILogger<GoGlobal> _logger;
 
        public string Source { get; set; } = ThirdParties.GOGLOBAL;

        public bool SupportsBookingSearch => true;

        public bool SupportsLiveCancellation(IThirdPartyAttributeSearch searchDetails, string source)
        {
            return _settings.AllowCancellations(searchDetails);
        }

        public bool SupportsRemarks => true;

        public int OffsetCancellationDays(IThirdPartyAttributeSearch searchDetails, string source)
        {
            return _settings.OffsetCancellationDays(searchDetails);
        }

        public bool RequiresVCard(VirtualCardInfo info, string source)
        {
            return false;
        }

        #endregion

        #region "Constructors"

        public GoGlobal(IGoGlobalSettings settings, ISerializer serializer, HttpClient httpClient, ILogger<GoGlobal> logger)
        {
            _settings = Ensure.IsNotNull(settings, nameof(settings));
            _serializer = Ensure.IsNotNull(serializer, nameof(serializer));
            _httpClient = Ensure.IsNotNull(httpClient, nameof(httpClient));
            _logger = Ensure.IsNotNull(logger, nameof(logger));
        }

        #endregion

        #region "Prebook"

        public async Task<bool> PreBookAsync(PropertyDetails oPropertyDetails)
        {
            bool bReturn = true;
            try
            {
                string[] searchReferences = oPropertyDetails.Rooms[0].ThirdPartyReference.Split('|');
                var valuationRequest = new PreBookRq
                {
                    HotelSearchCode = searchReferences[0],
                    ArrivalDate = oPropertyDetails.ArrivalDate.ToString(Constant.DataFormat)
                };
                string sRoot = await SendRequestAsync(oPropertyDetails, valuationRequest, Constant.PreBookRequestCode,
                    Constant.PreBookRequestNumber, "Pre-Book");

                oPropertyDetails.LocalCost = oPropertyDetails.Rooms.Sum(r => r.LocalCost);

                var prebookRs = _serializer.DeSerialize<Root<PreBookRs>>(sRoot).Main;

                if (!string.IsNullOrEmpty(prebookRs.Error))
                {
                    throw new Exception($"{prebookRs.Error}, {prebookRs.DebugError}");
                }

                if (!string.IsNullOrEmpty(prebookRs.Remarks))
                {
                    oPropertyDetails.Errata.AddNew("Important Information", prebookRs.Remarks);

                    var cancellations = GetCancellationPeriods(prebookRs.Remarks, oPropertyDetails.ArrivalDate)
                            .Select(c => new Cancellation
                            {
                                StartDate = c.StartDate,
                                EndDate = c.EndDate.AddDays(-1),
                                Amount = oPropertyDetails.LocalCost * c.FeePercent / 100
                            });

                    oPropertyDetails.Cancellations.AddRange(cancellations);
                }
            }
            catch (Exception ex)
            {
                oPropertyDetails.Warnings.AddNew("PreBook Exception", ex.Message);
                bReturn = false;
            }
            return bReturn;
        }

        #endregion

        #region "Book"

        public async Task<string> BookAsync(PropertyDetails oPropertyDetails)
        {
            string sReference;
            try
            {
                var bookReservationRequest = BookReservationRequest(oPropertyDetails);
                string xBook = await SendRequestAsync(oPropertyDetails, bookReservationRequest, Constant.BookRequestCode, Constant.BookRequestNumber, "Book");
                var bookInsertRs = _serializer.DeSerialize<Root<BookInsertRs>>(xBook).Main;

                if (!string.IsNullOrEmpty(bookInsertRs.Error))
                {
                    throw new Exception($"{bookInsertRs.Error}, {bookInsertRs.DebugError}");
                }

                sReference = bookInsertRs.GoBookingCode;

                var statusRequest = new BookStatusRq
                {
                    GoBookingCode = sReference
                };

                string xStatus = await SendRequestAsync(oPropertyDetails, statusRequest, Constant.StatusConfirmationRequestCode,
                                                            Constant.StatusConfirmationRequestNumber, "GoBookingCode");
                var bookStatusRs = _serializer.DeSerialize<Root<BookStatusRs>>(xStatus).Main;

                if (!string.IsNullOrEmpty(bookStatusRs.Error))
                {
                    throw new Exception($"{bookStatusRs.Error}, {bookStatusRs.DebugError}");
                }

                var bookingConfirmedStatuses = new List<string> { "C", "VCH" };
                if (!bookingConfirmedStatuses.Contains(bookStatusRs.GoBookingCode.Status))
                {
                    sReference = "Failed";
                }
            }
            catch (Exception ex)
            {
                oPropertyDetails.Warnings.AddNew("Book Exception", ex.Message);
                sReference = "Failed";
            }
            return sReference;
        }

        #endregion

        #region "Cancel"

        public async Task<ThirdPartyCancellationFeeResult> GetCancellationCostAsync(PropertyDetails oPropertyDetails)
        {
            var oThirdPartyCancellationFeeResult = new ThirdPartyCancellationFeeResult
            {
                Success = false
            };

            try
            {
                var oBookingSearchRq = new BookingSearchRq
                {
                    GoBookingCode = oPropertyDetails.SourceReference
                };

                string sBookingSeaqrchRs = await SendRequestAsync(oPropertyDetails, oBookingSearchRq, Constant.BookingSearchRequestCode,
                                                        Constant.BookingSearchRequestNumber, "Cancellation Cost");

                var xBookingSeaqrchRs = _serializer.DeSerialize<Root<BookingSearchRs>>(sBookingSeaqrchRs).Main;

                if (!string.IsNullOrEmpty(xBookingSeaqrchRs.Error))
                {
                    throw new Exception($"{xBookingSeaqrchRs.Error}, {xBookingSeaqrchRs.DebugError}");
                }

                var arrivalDate = ParseDate(xBookingSeaqrchRs.ArrivalDate, "yyyy-MM-dd");
                decimal totalPrice = xBookingSeaqrchRs.TotalPrice.ToSafeDecimal();
                string currencyCode = xBookingSeaqrchRs.Currency;
                var cDeadline = ParseDate(xBookingSeaqrchRs.CancellationDeadline, "yyyy-MM-dd");
                var now = DateTime.Now;
                var cancelDeadlineIncluded = new DateTime(cDeadline.Year, cDeadline.Month, cDeadline.Day).AddDays(1);

                if (now < cancelDeadlineIncluded) return new ThirdPartyCancellationFeeResult
                {
                    Amount = 0.00M,
                    CurrencyCode = currencyCode,
                    Success = true
                };

                var cancellationPeriods = GetCancellationPeriods(xBookingSeaqrchRs.Remark, arrivalDate);
                var currentPeriod = cancellationPeriods.FirstOrDefault(cp => cp.StartDate < now && cp.EndDate > now);

                if (currentPeriod != null)
                {
                    oThirdPartyCancellationFeeResult.Success = true;
                    oThirdPartyCancellationFeeResult.CurrencyCode = currencyCode;
                    oThirdPartyCancellationFeeResult.Amount = currentPeriod.FeePercent * totalPrice;
                }
            }
            catch (Exception ex)
            {
                oPropertyDetails.Warnings.AddNew("Cancellation Cost Exception", ex.Message);
                oThirdPartyCancellationFeeResult.Success = false;
            }

            return oThirdPartyCancellationFeeResult;
        }

        public async Task<ThirdPartyCancellationResponse> CancelBookingAsync(PropertyDetails oPropertyDetails)
        {
            var oTPCancellationResponse = new ThirdPartyCancellationResponse
            {
                Success = false,
                TPCancellationReference = "Failed"
            };

            try
            {
                var cancelRequst = new CancelRq
                {
                    GoBookingCode = oPropertyDetails.SourceReference
                };

                string xCancelRs = await SendRequestAsync(oPropertyDetails, cancelRequst, Constant.CancelRequestCode, Constant.CancelRequestNumber, "Cancel");
                var oCancelRs = _serializer.DeSerialize<Root<CancelRs>>(xCancelRs).Main;

                if (!string.IsNullOrEmpty(oCancelRs.Error))
                {
                    throw new Exception($"{oCancelRs.Error}, {oCancelRs.DebugError}");
                }

                var cancelConfirmedStatuses = new List<string> { "X", "XF" };
                if (cancelConfirmedStatuses.Contains(oCancelRs.BookingStatus))
                {
                    oTPCancellationResponse.TPCancellationReference = oCancelRs.GoBookingCode;
                    oTPCancellationResponse.Success = true;
                }
            }
            catch (Exception ex)
            {
                oPropertyDetails.Warnings.AddNew("Cancellation Exception", ex.Message);
                oTPCancellationResponse.TPCancellationReference = "Failed";
                oTPCancellationResponse.Success = false;
            }

            return oTPCancellationResponse;
        }

        #endregion

        #region "BookingSearch"

        public ThirdPartyBookingSearchResults BookingSearch(BookingSearchDetails bookingSearchDetails)
        {
            return new();
        }

        #endregion

        #region "Other Methods"

        public string CreateReconciliationReference(string inputReference)
        {
            return "";
        }

        public void EndSession(PropertyDetails propertyDetails)
        {
        }

        #endregion

        #region "Booking Status Update"
        public ThirdPartyBookingStatusUpdateResult BookingStatusUpdate(PropertyDetails propertyDetails)
        {
            return new();
        }

        #endregion

        #region "Helpers"

        internal async Task<string> SendRequestAsync<T>(PropertyDetails oPropertyDetails, T request, string operation,
                                        int operationCode, string logFileName) where T : Main, new()
        {
            var oRoot = new Root<T>
            {
                Header =
                {
                         Agency = _settings.Agency(oPropertyDetails),
                         User = _settings.User(oPropertyDetails),
                         Password = _settings.Password(oPropertyDetails),
                         Operation = operation,
                         OperationType = Constant.OperationTypeRequest,
                },
                Main = request
            };

            string sRoot = $"{_serializer.Serialize(oRoot).OuterXml}";
            var envelope = Envelope.CreateRequest(operationCode, sRoot, _serializer);

            //'Make the web request
            var oRequest = new Request
            {
                EndPoint = _settings.GenericURL(oPropertyDetails),
                Method = RequestMethod.POST,
                Source = Source,
                LogFileName = logFileName,
                CreateLog = true,
                UseGZip = true
            };
            oRequest.SetRequest(envelope);
            await oRequest.Send(_httpClient, _logger);

            oPropertyDetails.AddLog(logFileName, oRequest);

            string responseRoot = Envelope.GetResponse(oRequest.ResponseXML, _serializer);

            return responseRoot;
        }

        internal static BookInsertRq BookReservationRequest(PropertyDetails oPropertyDetails)
        {
            var insertRequest = new BookInsertRq
            {
                Version = Constant.Version_2_3,
                AgentReference = oPropertyDetails.BookingReference,
                HotelSearchCode = oPropertyDetails.Rooms[0].ThirdPartyReference.Split('|')[0],
                ArrivalDate = oPropertyDetails.ArrivalDate.ToString(Constant.DataFormat),
                Nights = oPropertyDetails.Duration,
                NoAlternativeHotel = "1",
                Remark = oPropertyDetails.Rooms[0].SpecialRequest,
                Leader =
                    {
                        LeaderPersonID = 1
                    },
                Rooms = oPropertyDetails.Rooms.Select(oRoom =>
                {
                    var adults = oRoom.Passengers.Where(p => p.PassengerType == PassengerType.Adult)
                        .Select((oGuest, idx) => new PersonName
                        {
                            FirstName = oGuest.FirstName,
                            LastName = oGuest.LastName,
                            Title = $"{oGuest.Title.ToUpper()}.",
                            PersonID = $"{idx + 1}"
                        }).ToList();

                    var childs = oRoom.Passengers.Where(p => p.PassengerType is PassengerType.Child or PassengerType.Infant)
                        .Select((oGuest, idx) => new ExtraBed
                        {
                            FirstName = oGuest.FirstName,
                            LastName = oGuest.LastName,
                            PersonID = $"{idx + 1 + adults.Count}",
                            ChildAge = oGuest.PassengerType == PassengerType.Child
                                       ? oGuest.Age
                                       : 1
                        }).ToList();

                    return new RoomType
                    {
                        Adults = adults.Count,
                        Room =
                        {
                            RoomID = oRoom.PropertyRoomBookingID,
                            Guests = adults,
                            ExtraBeds = childs
                        }
                    };
                }).ToList()
            };
            return insertRequest;
        }

        internal static List<CancelPeriod> GetCancellationPeriods(string remark, DateTime arrivalDate)
        {
            var cancellationPeriods = new List<CancelPeriod>();
            var cancellationRemark = ParseRemark(remark);
            if (cancellationRemark.Any())
            {
                int cxlsCount = cancellationRemark.Count;
                var cxls = cancellationRemark.Select(c => new { StartDate = c.Key, FeePercent = c.Value }).OrderBy(c => c.StartDate).ToList();

                cancellationPeriods = Enumerable.Range(0, cxlsCount).Select(i => new CancelPeriod
                {
                    StartDate = cxls[i].StartDate,
                    EndDate = (i + 1 < cxlsCount)
                            ? cxls[i + 1].StartDate
                            : arrivalDate,
                    FeePercent = cxls[i].FeePercent
                }).ToList();
            }

            return cancellationPeriods;
        }

        private static Dictionary<DateTime, decimal> ParseRemark(string remark)
        {
            var cancellations = new Dictionary<DateTime, decimal>();
            string template = @".*CXL charges apply as follows:(.*)";
            var matchRemark = Regex.Match(remark, template);
            bool hasCancellations = matchRemark.Success;
            if (hasCancellations)
            {
                remark = matchRemark.Groups[1].Value;
                string cancellationTemplate = @"STARTING (\d{2}/\d{2}/\d{4}) CXL-PENALTY FEE IS (\d+.?\d*)% OF BOOKING PRICE.";

                while (hasCancellations)
                {
                    var matchCancel = Regex.Match(remark, cancellationTemplate);
                    if (matchCancel.Success)
                    {
                        remark = remark.Replace(matchCancel.Groups[0].Value, "");
                        var startDate = ParseDate(matchCancel.Groups[1].Value, "dd/MM/yyyy");
                        decimal percent = matchCancel.Groups[2].Value.ToSafeDecimal();
                        cancellations.Add(startDate, percent);
                    }
                    else
                    {
                        hasCancellations = false;
                    }
                };
            }

            return cancellations;
        }

        public static DateTime ParseDate(string sDate, string format)
        {
            DateTime.TryParseExact(sDate, format, CultureInfo.InvariantCulture, DateTimeStyles.None, out var dDate);
            return dDate;
        }

        #endregion
    }
}
